### DISCUSSION TITLE
Disabling `snacks.toggle` keymaps

### DISCUSSION THREAD

**Discussion by ernstwi** [LazyVim/LazyVim#5240] [Q&A]
I think the default mappings set up for snacks.toggle cannot be cleanly disabled.

Example from `lua/lazyvim/config/keymaps.lua`:

```
Snacks.toggle.animate():map("<leader>ua")
```

This function sets up the mapping and also calls `require("which-key").add`.

The mapping can be removed using `vim.keymap.del` as suggested [here](https://www.lazyvim.org/configuration/keymaps#global-keymaps). However the which-key config for this mapping remains, and I don't see any function in the which-key API which allows me to manually undo the add.

I would expect either:
1. When a mapping is removed with `vim.keymap.del`, this also removes which-key settings for that mapping (I think this is the case with the other mappings in `keymaps.lua`?)
2. Mappings for snacks.toggle are set up using the `keys` property in the snacks plugin spec (allowing us to disable by setting a key to `false`).

**Comment by dpetka2001** [comment]
Yes, that's how it currently works it seems. I tried to rectify this behavior with [this PR in snacks repo](https://github.com/folke/snacks.nvim/pull/378).

**Comment by danjessen** [comment]
I currently have tabline and animations disabled, but LazyVim still sets up ALL the keymaps for disabled plugins still exist. That is a bit frustrating.

**Comment by windowsrefund** [comment]
bumping this as it does seem to be impossible to reuse `<leader>ua`. I can disable it with a `vim.keymap.del('n', '<leader>ua')` but once I go to use it again, I end up with the "Enable Animations" menu item in my which-key menu.

I don't make any claim to understand the nitty details but shouldn't these Snack-dependent keymaps be maintained in the `keys` element of the Snacks lazy config? It just seems like the global [keymaps.lua](https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/config/keymaps.lua) is the wrong place. 

It looks like "something" was done [here](https://github.com/folke/snacks.nvim/commit/ee9e6179fe18a2bf36ebb5e81ddf1052e04577dc) but I have no idea what "hide toggle when real keymap does not exist." could possibly mean. Can someone explain that in English and provide a clear example of how to proceed?

**Reply by dpetka2001** [reply]
The easiest way to replace a toggle keymap would be to add in your `~/.config/nvim/lua/config/keymaps.lua` the following
```lua
require("which-key").add({
  { "<leader>ua", "<cmd>echo 'test'<cr>", desc = "test", mode = "n" },
  -- you can add further keymaps here to replace
})
```
Otherwise if you just want to disable it, you just do `vim.keymap.del("n", "<leader>ua")` and it won't show.

**Reply by windowsrefund** [reply]
aha! I didn't think to lean into which-key key like that. Will give that approach a go. That said, I'm still going to stand by my original suspicion this is really disjointed. There's a sense on my end that sometimes you can/should overwrite a binding in that file. Other times (like here) you can't and have to reach out to which-key directly. Then, there's the idea you'd want to tie a binding to a particular plugin via that individual `keys` element so the mapping only exists when a plugin is used. That latter approach doesn't work here (but does in other cases). So again, it just seems like the snacks mappings are being handled/managed as a special snowflake which only makes things more (needlessly?) complicated.

**Reply by windowsrefund** [reply]
Right, as mentioned above, I've just confirmed this approach does not work from within a `config` for a given plugin. As suggested, it DOES work when called from the global `keymaps.lua` file but that's a poor pattern to follow as we're now assuming the required plugin is installed. Sure, we can add checking bloat around it but that kinda reinvents the wheel for no good reason.

**Reply by dpetka2001** [reply]
Please give an example of the plugin spec you're trying to implement.

**Reply by windowsrefund** [reply]
Why? That seems like a needless distraction as the points I've raised are plugin-agnostic.

**Reply by dpetka2001** [reply]
If you load your plugin like the following it will work
```lua
{
    "kylechui/nvim-surround",
    event = "LazyFile",
    opts = function()
      require("which-key").add({
        { "<leader>ua", "<cmd>echo 'test'<cr>", desc = "test", mode = "n" },
      })
    end,
  }
```
For me it changes the description and shows a message notification "test". `event = "LazyFile"` is needed so the plugin will load after `keymaps.lua`, thus overwriting the default keymap. 

And if the plugin is not installed, then the default mapping will take effect.

**Reply by windowsrefund** [reply]
OK, I guess we're going to get into the nitty details on a plugin-specific basis. This does sorta work but only after a file is loaded into the buffer. If I just run Neovim with nothing loaded and then look at the menu, the original mapping and description is present (unless I also disable it in the global `keymaps.lua` file).

```
  {
    "altermo/ultimate-autopair.nvim",
    enabled = true,
    -- event = { "InsertEnter", "CmdlineEnter", "LazyFile" },
    event = "LazyFile",
    branch = "v0.6", --recommended as each new version will have breaking changes
    config = function()
      require("ultimate-autopair").setup(ua_opts)

      local wk = require("which-key")
      wk.add({
        {
          "<leader>ua",
          function()
            require("ultimate-autopair").toggle()
          end,
          desc = "Toggle autopair",
          mode = "n",
        },
      })
    end,
  },
```

While I'm sure this can be "made to work", aren't we just dancing around tthe original issue you yourself observed with the clobbering effect of the global `<leader>ua` (and other friends) mappings?

Why aren't all the snacks-specific keymaps just maintained in the snacks spec so we don't have to work this hard on a case-by-case basis?

**Reply by dpetka2001** [reply]
Because these mappings are not snacks-specific. 

In plugin specs you usually find mappings that have to do with specific plugin functionality (for example in Telescope you find Telescope specific mappings, in FzfLua the same and Snacks.picker also the same). Toggling animations, spelling, wrapping, line numbers etc are just general Neovim functionalities that just use `Snacks.toggle` to define the mappings, so that they are dynamically defined in combination with which-key. At least that's how I understand it. Also, you are right that you still have to disable the global mapping in `keymaps.lua` if you don't want it to appear when you open Neovim and are just looking at the dashboard for example until a buffer loads.

Also, I'm just a simple user like you and the maintainer makes the decisions for this project.

**Reply by windowsrefund** [reply]
[These mappings](https://github.com/LazyVim/LazyVim/blob/ec5981dfb1222c3bf246d9bcaa713d5cfa486fbd/lua/lazyvim/config/keymaps.lua#L139) are not snacks-specific? LOL, that was absolutely rhetorical of course. Anyway, I keep trying to bring the focus back to the anti-pattern with regard to the well-established `keys` element of the Lazy spec. It's like we're willingly fighting against ourselves here because we have 2 opposing workflows in place.

**Reply by dpetka2001** [reply]
Just do a [search](https://github.com/search?q=repo%3ALazyVim%2FLazyVim%20%22Snacks.toggle%22&type=code) and you'll see other mappings defined in the plugin specs themselves (mini.animate, mini.pairs, treesitter-context etc), because they offer a functionality specific to those plugins. The mappings you linked are general Neovim functionality. 

If we see things different, then we'll just have to agree to disagree on that. Nothing wrong about it.

And if you have such a request you should point it (I'm not sure if that's valid English, so excuse me about the expression) towards the maintainer himself, as like I said he's the one making the decisions.

**Reply by windowsrefund** [reply]
With all due respect, I don't know why I need to say everything 3 or more times. I link to snack-specific global keymaps and you respond with "the mappings are not snacks-specific" and then double down with "the mappings you linked are general Neovim functionality"....

What? :)

Also, I wasn't pointing this observation to anyone in particular. All I did was jump onto an existing thread and supply additional information and context to a problem that was already known (you yourself tried to solve it). 

It's just a little frustrating to point out something so obvious (the sky is blue) and now we're in the weeds with work-arounds that "kinda" work but are needlessly heavy. Hardly anyone is going to know how to overcome the original issue so they'll have to dig through these issues; hopefully they'll come across this and kinda/sorta benefit.

When can we just say "yea, there's a anti-pattern here and it kinda sucks?"

**Comment by windowsrefund** [comment]
Figured I'd share what I landed on as it gets the job done in a way that minimizes the assumption that the targeted plugin is available. My `lua/config/keymaps.lua` file now contains the following bits:

```
wk = require("which-key")
wk.add({
  {
    "<leader>ua",
    function()
      local notify = vim.notify
      local function bool2str(bool)
        return bool and "on" or "off"
      end
      local ok, ua = pcall(require, "ultimate-autopair")
      if ok then
        ua.toggle()
        vim.g.ua_enabled = require("ultimate-autopair.core").disable
        notify(string.format("ultimate-autopair %s", bool2str(not vim.g.ua_enabled)))
      else
        notify("ultimate-autopair not available")
      end
    end,
    desc = "Toggle autopair",
    mode = "n",
  },
})
```

Thanks again for the discussion @dpetka2001
